### ISSUE TITLE
Support task pool shutdowns on idle

### ISSUE THREAD

**Issue by smerkousdavid** [celery/celery#7758]
# Checklist

- [x] I have checked the [issues list](https://github.com/celery/celery/issues?utf8=%E2%9C%93&q=is%3Aissue+label%3A%22Issue+Type%3A+Feature+Request%22+)
  for similar or identical feature requests.
- [x] I have checked the [pull requests list](https://github.com/celery/celery/pulls?utf8=%E2%9C%93&q=is%3Apr+label%3A%22PR+Type%3A+Feature%22+)
  for existing proposed implementations of this feature.
- [x] I have checked the [commit log](https://github.com/celery/celery/commits/master)
  to find out if the same feature was already implemented in the
  master branch.
- [x] I have included all related issues and possible duplicate issues
  in this issue (If there are none, check this box anyway).

#### Related Issues

- Could not find anything for the key terms I searched

#### Possible Duplicates

- This seems similar to where a celery worker could be "bursted" and shutdown once the queue is empty here #6687. Would be a similar feature

# Brief Summary

Add the ability to shutdown a worker/pool, not entirely knowledgeable of underlying processes, when a queue is empty. I currently have the issue with tensorflow not deallocating GPU and the only solution would be to use multiprocessing, which I presume celery uses under the hood for certain pools, and it seems janky to have a task start another process, which just complicates things. Thus a method of "restarting" the pool after no tasks have been submitted for X amount of time would be nice.

# Design

## Architectural Considerations

None

## Proposed Behavior

It would be nice to start a celery worker have it process whatever is in the queue and once it's done close/restart the pool process so that any allocated memory would be cleared. This is sort of similar to max-tasks-per-child, but that I noticed only restarts when a NEW task is assigned not when it's idle.

Some logic that would make this work better is that the pool wouldn't restart unless if it has already processed some tasks. In my case the queue could be empty majority of the time, but I occasionally have some peak GPU load for site tensorflow stuff and would like the memory to be cleared once the queue sits empty for let's say 5 minutes. I've looked at tf issues and it seems like there really isn't any non-obscure way to clear the memory without ending the process/clearing cuda/resetting the device which is not what I want. 

## Proposed UI/UX

Add a flag of the sort, possibly way better names, `--child-empty-timeout Xs` which would reset the worker after X seconds of idle AFTER processing something.

## Diagrams

N/A

## Alternatives

Simply have the `max-tasks-per-child` terminate and recreate the pool immediately after the N tasks is complete.

**Comment by auvipy** [comment]
do you have implementation plan? though Im not 100% sure about the design decision/implementation detail yet

**Comment by thedrow** [comment]
@smerkousdavid Is there a memory leak when worker processes are idle?
I'm open to accepting a contribution for such a feature but I'm not sure about the use-case yet.

**Comment by smerkousdavid** [comment]
@auvipy I don't have an implementation plan yet and might have more time this summer to check out the celery codebase for that.

@thedrow Sort of. Tensorflow isn't too friendly about releasing memory, and at the time of writing the issue (probably still is), it is all managed by the c++ backend, which aren't deallocated until the python process is stopped. I wouldn't call it a memory leak from tf, but it does cause issues when I have an idle celery task and 20GB of my GPU memory is consumed unless if I restart celery itself anytime another process (other than celery) for training wants to use the GPU. I just need to figure out some feature that will kill the worker after a certain idle time.

**Comment by kaedroho** [comment]
I think this functionality would be useful for saving costs when running Celery workers on scale-to-zero platforms like Fly.io, Google Cloud Run, and Azure Container Apps.

For example, see how this can be done with Laravel on Fly.io: https://fly.io/laravel-bytes/cost-effective-queue-workers-with-fly-io-machines/

**Comment by corinwagen** [comment]
Yeah, I agree w/ @kaedroho that this could be useful. Sometimes it's nice to have mixed-backend queues where you can add serverless workers to meet demand spikes while keeping long-running workers alive on dedicated hardware.

**Comment by ajoy39** [comment]
@kaedroho Here I am, a month after your post, trying to figure out how to do just this with Google Cloud Run Jobs lol. Currently trying to create a custom Task class that checks the queue for active or reserved jobs after_return and shuts the workers down if they're both empty (after filtering the current job out anyway). Unfortunately running into a race condition at this point where the app shuts down before the last worker runs after_return and control() is no longer defined.
